src/fiber/history.rs

```diff
@@ -0,0 +1,422 @@
+    pub min_fail_relax_interval: u128,
+    pub bimodal_scale_msat: f64,
+    // this filed is used to check whether from is the source Node
+    // will be used after enabling the direct channel related logic
+    #[allow(dead_code)]
+    pub source: Pubkey,
+    store: S,
+}
+
+impl<S> PaymentHistory<S>
+where
+    S: NetworkGraphStateStore + Clone + Send + Sync + 'static,
+{
+    pub(crate) fn new(source: Pubkey, min_fail_relax_interval: Option<u128>, store: S) -> Self {
+        let mut s = PaymentHistory {
+            source,
+            inner: HashMap::new(),
+            min_fail_relax_interval: min_fail_relax_interval
+                .unwrap_or(DEFAULT_MIN_FAIL_RELAX_INTERVAL),
+            bimodal_scale_msat: DEFAULT_BIMODAL_SCALE_SHANNONS,
+            store,
+        };
+        s.load_from_store();
+        s
+    }
+
+    #[cfg(test)]
+    pub(crate) fn reset(&mut self) {
+        self.inner.clear();
+    }
+
+    pub(crate) fn add_result(&mut self, from: Pubkey, target: Pubkey, result: TimedResult) {
+        self.inner
+            .entry(from)
+            .or_insert_with(HashMap::new)
+            .insert(target, result);
+        self.save_result(from, target, result);
+    }
+
+    fn save_result(&mut self, from: Pubkey, target: Pubkey, result: TimedResult) {
+        self.store
+            .insert_payment_history_result(from, target, result);
+    }
+
+    pub(crate) fn load_from_store(&mut self) {
+        let results = self.store.get_payment_history_result();
+        for (from, target, result) in results.iter() {
+            self.inner
+                .entry(from.clone())
+                .or_insert_with(HashMap::new)
+                .insert(target.clone(), *result);
+        }
+    }
+
+    pub(crate) fn apply_pair_result(
+        &mut self,
+        from: Pubkey,
+        target: Pubkey,
+        amount: u128,
+        success: bool,
+        time: u128,
+    ) {
+        let min_fail_relax_interval = self.min_fail_relax_interval;
+        let result = if let Some(current) = self.get_mut_result(&from, &target) {
+            if success {
+                current.success_time = time;
+                if amount > current.success_amount {
+                    current.success_amount = amount;
+                }
+                if current.fail_time != 0 {
+                    current.fail_amount = amount + 1;
+                }
+            } else {
+                if amount > current.fail_amount
+                    && current.fail_time != 0
+                    && time.saturating_sub(current.fail_time) < min_fail_relax_interval
+                {
+                    return;
+                }
+                current.fail_amount = amount;
+                current.fail_time = time;
+                if amount == 0 {
+                    current.success_amount = 0;
+                } else if amount <= current.success_amount {
+                    current.success_amount = amount - 1;
+                }
```

Comment:
the condition `amount == 0` implies `amount <= current.success_amount`. 

```suggestion
if amount <= current.success_amount {
                    current.success_amount = amount.saturating_sub(1);
}
```